src/inc/corprof.idl

```diff
@@ -2375,6 +2375,34 @@ interface ICorProfilerCallback7 : ICorProfilerCallback6
 }
 
 
+[
+    object,
+    uuid(5BED9B15-C079-4D47-BFE2-215A140C07E0),
+    pointer_default(unique),
+    local
+]
+interface ICorProfilerCallback8 : ICorProfilerCallback7
+{
+    // This event is triggered whenever a metadata less method is jit compiled.
+    // These include IL Stubs, DynamicMethod or more generally any use of LCG.
+    // The goal is to provide profiler writers with enough information to identify
+    // it to users as beyond unknown code addresses.
+    // Note: FunctionID's provided here cannot be used to resolve to their metadata
+    //       tokens since LCG methods have no metadata.
+    HRESULT LightweightJITCompilationStarted(
+        FunctionID functionId,
+        UINT_PTR   ilHeader,
+        [in, string] const WCHAR * className,
+        [in, string] const WCHAR * methodName,
+        [in, string] const WCHAR * methodSignature);
+
```

Comment:
I agree, I wasn't sure of the bar for 463 (which is why the haste in doing this), but I think a reasonable API here is not difficult to achieve and shouldn't take me a while to redo.

How about something like this?

```
FunctionHasNoMetadata(/* [in]  */ FunctionID  functionId,
                      /* [out] */ BOOL        *hasNoMetadata);
```

```
GetNoMetadataFunctionInfo(/* [in]  */ FunctionID              functionId,
                          /* [out] */ ModuleID                *moduleId,
                          /* [out] */ PCCOR_SIGNATURE         *ppvSig,
                          /* [out] */ ULONG                   *pbSig,
                          /* [out] */ ULONG                   cchName,
                          /* [out] */ ULONG                   *pcchName,
 _Out_writes_to_(cchName, *pcchName)  WCHAR                   szName[]);
```

Seems like there can be more than one anonymously hosted assembly for lcg methods, so providing the module id in case it's useful for someone.

The name is the only string now and it needs to be there because there would be no other way to get to it, unless we add another API, though we can implement this part pay-for-play (if nullptr, don't give the result like all our other methods).

So we'll have these on ICorProfilerCallback:

Changing the names of the callbacks at your suggestion gives us this:

``JITCompilationStartedNoMetadata(FunctionID);``
``JITCompilationFinishedNoMetadata(FunctionID);``

And then finally one more addition to the callback I'd like to add is:

``PrecompiledMethodLoaded(FunctionID);``

This will provide information to the profiler about which NGEN and ReadyToRun methods have been encountered by the runtime. Loaded here in my opinion is a better word despite is not being strictly true.

The fundamental reason to add this API is now that R2R is here native module semantics have changed somewhat. One can search the address space of an NGEN image and be able to query the runtime (via GetFunctionFromIP) to get details about the EIPs the profiler may encounter.